File: src/queue.cpp
#include "queue.h"

volatile uint8_t q_first = 0;
volatile uint8_t q_last = 0;
can_message_t *queue_array[QUEUE_SIZE];

void queue_add(can_message_t *item) {
  queue_array[q_last] = item;
  q_last = (q_last + 1) % QUEUE_SIZE;
}

can_message_t *queue_remove() {
  if (q_first == q_last) {
    return nullptr;
  }
  auto *val = queue_array[q_first];
  q_first = (q_first + 1) % QUEUE_SIZE;
  return val;
}

uint8_t queue_empty() { return q_first == q_last; }

File: mission-control/src/queue.rs
const QUEUE_SIZE: u8 = 16;

pub struct Queue<T> {
    q_first: u8,
    q_last: u8,
    queue_array: [Option<T>; 16],
}

impl<T> Queue<T> {
    pub fn new() -> Self {
        return Queue {
            q_first: 0,
            q_last: 0,
            queue_array: std::array::from_fn(|_| None),
        };
    }

    pub fn add(&mut self, item: T) {
        self.queue_array[usize::from(self.q_last)] = Some(item);
        self.q_last = (self.q_last + 1) % QUEUE_SIZE;
    }

    pub fn remove(&mut self) -> Option<T> {
        if self.is_empty() {
            return None;
        }
        let item = self.queue_array[usize::from(self.q_first)].take();
        self.q_first = (self.q_first + 1) % QUEUE_SIZE;
        return item;
    }

    pub fn is_empty(&self) -> bool {
        return self.q_first == self.q_last;
    }
}
